src/serac/physics/utilities/functional/domain_integral_cuda.cuh

```diff
@@ -0,0 +1,259 @@
+
+  // for each element in the domain
+  const int grid_stride = blockDim.x * gridDim.x;
+  for (int e = blockIdx.x * blockDim.x + threadIdx.x; e < num_elements; e += grid_stride) {
+    // get the DOF values for this particular element
+    auto u_elem = detail::Load<trial_element>(u, e);
+
+    // this is where we will accumulate the element residual tensor
+    element_residual_type r_elem{};
+
+    // for each quadrature point in the element
+    for (int q = 0; q < static_cast<int>(rule.size()); q++) {
+      eval_quadrature<g, test, trial, Q, derivatives_type, lambda>(e, q, u_elem, r_elem, derivatives_ptr, J, X,
+                                                                   num_elements, qf);
+    }
+
+    // once we've finished the element integration loop, write our element residuals
+    // out to memory, to be later assembled into global residuals by mfem
+    detail::Add(r, r_elem, e);
+  }  // e loop
+}
+
+template <Geometry g, typename test, typename trial, int Q, typename derivatives_type, typename lambda,
+          typename solution_type, typename residual_type, typename jacobian_type, typename position_type>
+__global__ void eval_cuda_quadrature(const solution_type u, residual_type r, derivatives_type* derivatives_ptr,
+                                     jacobian_type J, position_type X, int num_elements, lambda qf)
+{
+  using test_element          = finite_element<g, test>;
+  using trial_element         = finite_element<g, trial>;
+  using element_residual_type = typename trial_element::residual_type;
+  static constexpr auto rule  = GaussQuadratureRule<g, Q>();
+
+  const int grid_stride = blockDim.x * gridDim.x;
+  // launch a thread for each quadrature x element point
+  for (int qe = blockIdx.x * blockDim.x + threadIdx.x; qe < num_elements * rule.size(); qe += grid_stride) {
+    // warps won't fetch that many elements ... not great.. but not horrible
+    int e = qe / rule.size();
+    int q = qe % rule.size();
+
+    // get the DOF values for this particular element
+    auto u_elem = detail::Load<trial_element>(u, e);
+
+    // this is where we will accumulate the element residual tensor
+    element_residual_type r_elem{};
+
+    // for each quadrature point in the element
+    eval_quadrature<g, test, trial, Q, derivatives_type, lambda>(e, q, u_elem, r_elem, derivatives_ptr, J, X,
+                                                                 num_elements, qf);
+
+    // once we've finished the element integration loop, write our element residuals
+    // out to memory, to be later assembled into global residuals by mfem
+    detail::Add(r, r_elem, e);
+  }  // quadrature x element loop
+}
+
+template <Geometry g, typename test, typename trial, int Q, serac::detail::ThreadExecutionPolicy policy,
+          typename derivatives_type, typename lambda>
+void evaluation_kernel_cuda(serac::detail::ThreadExecutionConfiguration config, const mfem::Vector& U, mfem::Vector& R,
```

Comment:
the consistency of naming on these function is slightly jarring... 

evaluation_kernal_cuda
eval_cuda_quadrature
gradient_cuda_element
gradient_kernel_cuda

Reply:
I discussed this with Jonathan last week, our plan was to make them into overloads of the 3 fundamental kernels
```cpp
evaluation_kernel< execution_space >
directional_derivative< execution_space >
gradient_matrix< execution_space >
```

Reply:
I say leave this for now and fix it in a follow up PR. I'll add a github issue.